backupccl: move conflicting system table descriptors to higher IDs

stevendanna opened the pull request:
This change fixes a cluster restore bug where backed up
non-system descriptors would clobber system tables in the restoring
cluster because of an ID collision.

As of 22.1 system tables no longer have reserved IDs, and can
be dynamically assigned IDs from 50 to 100. When performing a cluster
restore we enforce that the restoring cluster is free of any non-system
descriptors. We also incorrectly assumed that there will be no collision
between the backed up descriptors and the system descriptors in the
restoring clusters. These assumptions meant that we do not re-assign
any of the backed up descriptors, new IDs, but instead write them to the
restoring cluster with the same ID as when they were backed up.

In 22.1 we introduced two new system tables.  As of this change we have introduced
atleast another two system tables, and this trend is expected to continue.
Since restore is expected to support restoring a backup taken on an older
cluster version into a newer cluster version it is easy to construct a case
where IDs 50, 51 and so on were previously occupied by non-system table
descriptors. Concretely, in 21.2 ID 50 was occupied by `defaultdb`, and so
an attempt to restore a 21.2 backup into a 22.1+ cluster resulted in the
backed up `defaultdb` descriptor clobbering the newly introduced system table
at ID 50. Apart from having more validation around what a restore is allowed
to write to the `system.descriptors` and `system.namespace` table
(https://github.com/cockroachdb/cockroach/issues/84561), we need to ensure
that users are able to restore their older backups.

To this effect, we add another step to the restore planning phase that does
the following:

1. Collects all table descriptors in the restoring clusters `system` database.

2. Bumps the restoring clusters descriptor ID generator to a value higher than
any descriptor in the backup.

3. For every table in the `system` database that collides with a backed up
descriptor, restore will re-create the `system` table using the updated ID
generator. Thereby guaranteeing that it will not collide with any of the descriptors
we are about to restore.

4. Allow the cluster restore to progress as before.

Note, this is meant to serve as a relatively short term fix until we
desgin a better solution around cluster restores and dynamic system
tables. The only system tables that a cluster restore can conflict with
today are:

- `tenant_settings`: system tenant only
- `span_count`: non-system tenant only
- `privileges`: system and non-system tenant

Release note (bug fix): fixes a bug where cluster restores of older
backups would silently clobber system tables or fail to complete.

Addresses https://github.com/cockroachdb/cockroach/issues/84359

stevendanna commented:
Very very WIP.

stevendanna commented:
Questions:

- [ ] As a general strategy, what would we do about rangefeeds that have already been started on the spans related to the "old" system table.
- [ ] Do AOST queries involving these tables get weird?

stevendanna commented on pkg/ccl/backupccl/restore_planning.go:
```diff
@@ -1720,6 +1902,10 @@ func doRestorePlan(
 			return err
 		}
+
+		if err := moveConflictingSystemTables(ctx, p, databasesByID, schemasByID, filteredTablesByID, typesByID); err != nil {
```
This is a pretty heavy operation to have in planning, but for now I wanted to make sure this work is completely done before we start thinking about rewrites.

adityamaru commented on pkg/ccl/backupccl/restore_planning.go:
```diff
@@ -158,6 +159,237 @@ func synthesizePGTempSchema(
+		for _, sysTableDesc := range systemTableDescs {
+			id := sysTableDesc.GetID()
+			if desc, ok := databasesByID[id]; ok {
+				if desc.GetParentID() != systemschema.SystemDB.GetID() {
```
This should be GetID() instead of GetParentID() I think.

stevendanna replied:
Nice catch, fixed.

stevendanna commented on pkg/ccl/backupccl/restore_planning.go:
```diff
@@ -158,6 +159,237 @@ func synthesizePGTempSchema(
+		}
+
+		// Copy data from the existing table over to the temporary table.
+		insertStmt := fmt.Sprintf("INSERT INTO system.%s (SELECT * FROM system.%s)", newTable.Name, newTable.GetName())
```
Eek, a last minute refactor broke this. It should be oldTable.GetName()

stevendanna replied:
Fixed.

stevendanna commented on pkg/ccl/backupccl/restore_planning.go:
```diff
@@ -158,6 +159,237 @@ func synthesizePGTempSchema(
+			return err
+		}
+
+		// ClearRange on the table span for the old table that we've just migrated.
```
I added this ClearRange thinking it was prudent to clear out the keyspace we are about to potentially reuse. But the tests I added don't fail even in its absence.

stevendanna commented:
I think in practice we don't need to worry about it right now, but here we don't move the zone configurations or comments related to the tables we are moving.

We also theoretically have the same issue with schema changes on the system database, but I think there we would be better off just disallowing cluster restores if there are migrations still happening.

adityamaru commented on pkg/ccl/backupccl/restore_planning.go:
```diff
@@ -158,6 +159,251 @@ func synthesizePGTempSchema(
+		// move.
+		newTable := oldTable.NewBuilder().BuildCreatedMutable().(*tabledesc.Mutable)
+
+		// Allow the upgrades package to dynamically assign an ID to the new system
+		// table.
+		newTable.ID = descpb.InvalidID
+		newTable.Name = fmt.Sprintf("crdb_internal_copy_%s", newTable.Name)
+		id, created, err := upgrades.CreateSystemTableInTxn(ctx, p.ExecCfg().DB, txn, codec, newTable)
+		if err != nil {
+			return err
+		}
+		if !created {
+			return errors.New("temporary table for system table already exists")
+		}
+
+		// Copy data from the old table to the new table.
+		insertStmt := fmt.Sprintf("INSERT INTO system.%s (SELECT * FROM system.%s)", newTable.Name, oldTable.GetName())
+		if _, err := p.ExecCfg().InternalExecutor.Exec(ctx, "copy-to-temp-table", txn, insertStmt); err != nil {
+			return err
+		}
+
+		// Delete the old namespace entry for the new table.
+		b.Del(catalogkeys.EncodeNameKey(codec, newTable))
+
+		// Delete the descriptor entry for the old system table.
+		b.Del(catalogkeys.MakeDescMetadataKey(codec, oldTable.GetID()))
+		descsCol.AddDeletedDescriptor(oldTable.GetID())
+
+		// Overwrite the namespace entry for the system table with our new id, and update the
+		// descriptor of the new table with the old table's name.
+		newTable.ID = id
+		newTable.Name = oldTable.GetName()
+		if err := descsCol.WriteDescToBatch(ctx, false /* kvTrace */, newTable, b); err != nil {
+			return err
+		}
+		b.Put(catalogkeys.EncodeNameKey(codec, newTable), newTable.ID)
+
+		if err := txn.Run(ctx, b); err != nil {
+			return err
+		}
```
So I'm generating fixtures for older backups in 21.1 onwards and I'm getting a validation error which might be related to your comment about having to b.Put the new descriptor instead of going through WriteDescToBatch. This got me trying an alternate approach of:

- CREATE TABLE copy AS SELECT * from old_system;
- DROP TABLE old_system;
- ALTER TABLE copy RENAME TO old_system;

instead of this KV level manipulation. This does mean that we have a dropped descriptor waiting out GC, but is this a big deal? We already drop defaultdb and postgres as a part of restore anyways.

stevendanna replied:
@adityamaru I like the idea of doing it all via SQL.

stevendanna replied:
> This does mean that we have a dropped descriptor waiting out GC, but is this a big deal?

Will we also have a GCJob waiting around that is going to clear out the table keyspace?  Not a blocker, just something to double check when we take a look.

ajwerner replied:
Can we special case the logic that runs here to behave differently for descriptors in the system database?

adityamaru replied:
Good idea, that's what I ended up doing.

adityamaru committed:
backupccl: move conflicting system tables to higher IDs

This change fixes a cluster restore bug where backed up
non-system descriptors would clobber system tables in the restoring
cluster because of an ID collision.

As of 22.1 system tables no longer have reserved IDs, and can
be dynamically assigned IDs from 50 to 100. When performing a cluster
restore we enforce that the restoring cluster is free of any non-system
descriptors. We also incorrectly assumed that there will be no collision
between the backed up descriptors and the system descriptors in the
restoring clusters. These assumptions meant that we do not re-assign
any of the backed up descriptors, new IDs, but instead write them to the
restoring cluster with the same ID as when they were backed up.

In 22.1 we introduced two new system tables.  As of this change we have introduced
atleast another two system tables, and this trend is expected to continue.
Since restore is expected to support restoring a backup taken on an older
cluster version into a newer cluster version it is easy to construct a case
where IDs 50, 51 and so on were previously occupied by non-system table
descriptors. Concretely, in 21.2 ID 50 was occupied by `defaultdb`, and so
an attempt to restore a 21.2 backup into a 22.1+ cluster resulted in the
backed up `defaultdb` descriptor clobbering the newly introduced system table
at ID 50. Apart from having more validation around what a restore is allowed
to write to the `system.descriptors` and `system.namespace` table
(https://github.com/cockroachdb/cockroach/issues/84561), we need to ensure
that users are able to restore their older backups.

To this effect, we add another step to the restore planning phase that does
the following:

1. Collects all table descriptors in the restoring clusters `system` database.

2. Bumps the restoring clusters descriptor ID generator to a value higher than
any descriptor in the backup.

3. For every table in the `system` database that collides with a backed up
descriptor, restore will re-create the `system` table using the updated ID
generator. Thereby guaranteeing that it will not collide with any of the descriptors
we are about to restore.

4. Allow the cluster restore to progress as before.

Note, this is meant to serve as a relatively short term fix until we
desgin a better solution around cluster restores and dynamic system
tables. The only system tables that a cluster restore can conflict with
today are:

- `tenant_settings`: system tenant only
- `span_count`: non-system tenant only
- `privileges`: system and non-system tenant

Release note (bug fix): fixes a bug where cluster restores of older
backups would silently clobber system tables or fail to complete.

adityamaru commented:
I think this is ready for a proper review now. Following is the consolidated list of known shortcoming of this approach that I'd like to signoff / get comments on:

- [ ] We don't do anything about the rangefeed on `tenant_settings` when moving the table to a higher ID. Is this a problem?
- [ ] We don't handle the case where a cluster restore and a cluster version update run concurrently. A system table created as part of migration could show up somewhere in the restore after we've checked for conflict. Should we do something about this?
- [ ] Zone configs, comments and any other system table that has an ID link to the old system table is not being updated to the new ID of the system table. Is this a problem?

ajwerner commented:
> We don't do anything about the rangefeed on tenant_settings when moving the table to a higher ID. Is this a problem?

I think this one is a problem. I think It means that you'll have to restart the nodes after a restore before cluster settings will work for secondary tenants. I'm not exactly sure what to do here, but I have a feeling we can do something without that much effort. I'd be keen to look into some method we could invoke to reinitialize the setting watching subsystem. I *think* we have the right tools to make this happen inside each node, perhaps via some callback we plumb into the lease manager. 

>  We don't handle the case where a cluster restore and a cluster version update run concurrently. A system table created as part of migration could show up somewhere in the restore after we've checked for conflict. Should we do something about this?

I think this is fine. We don't support a full cluster restore with a non-empty cluster. I think it's fine if we also say that you can't upgrade during the restore.

>  Zone configs, comments and any other system table that has an ID link to the old system table is not being updated to the new ID of the system table. Is this a problem?

I suspect this is okay. We could do something here if we had to. My 2c is let's not solve this problem for these two tables right now, but let's make sure that whatever we end up doing in the long term doesn't have this problem.

RichardJCai commented on pkg/sql/catalog/catprivilege/validate.go:
```diff
@@ -54,6 +55,12 @@ func allowedSuperuserPrivileges(objectNameKey catalog.NameKey) privilege.List {
 		return privs
 	}
+	// If the object is not a registered system table but is in the system
+	// database then return the same privileges a registered system table would
```
nit: this is true right now since we're only doing it for one table but if for whatever reason we had to do this for another table that was `ReadOnly` this would be inaccurate.

Maybe it'll be better to special case it the key lookup and do 
			`tableKey.Name = table we actually want to looukp` ?

adityamaru replied:
So if IIUC before this diff all tables in the system database that weren't registered system tables were getting `ALL` anyways so we're no worse off with this diff when it comes to `ReadOnly` tables? We could special case for just `crdb_internal_copy_tenant_settings` and `crdb_internal_copy_span_count` but that just seemed more brittle.

adityamaru replied:
Though maybe the special casing prevents all the test churn below as well 🤷‍♂️

RichardJCai replied:
>So if IIUC before this diff all tables in the system database that weren't registered system tables were getting ALL anyways so we're no worse off with this diff when it comes to ReadOnly tables? We could special case for just crdb_internal_copy_tenant_settings and crdb_internal_copy_span_count but that just seemed more brittle.

Fair enough, I'm still inclined to go with the special casing for the above two and defining the two `crdb_internal_...` system tables. 

I don't think it's necessarily brittle and I'm not sure we really want the catch all case here.

adityamaru replied:
Added special case for only tables in the system database and with a table prefix equal to our sentinel prefix for a copied system table.

RichardJCai commented on pkg/sql/catalog/tabledesc/validate_test.go:
```diff
@@ -1977,7 +1977,7 @@ func TestValidateTableDesc(t *testing.T) {
 	for i, d := range testData {
 		t.Run(d.err, func(t *testing.T) {
-			d.desc.Privileges = catpb.NewBasePrivilegeDescriptor(username.RootUserName())
+			d.desc.Privileges = catpb.NewReadWriteSystemTablePrivilegeDescriptor(username.RootUserName())
```
Should we update these tables to not be part of the system table and create a separate test for system tables here?

adityamaru replied:
I did start on that path but the diff was a lot larger so I switched to this more minimal diff. I figured the Privileges we were setting here weren't the crux of what we were testing so it didn't matter much. I'm happy to go through the motions of bumping all tableIDs to 2 and their parents to 3 or something if others feel strongly about it.

adityamaru replied:
This is moot now that we only special case for the system tables created by the restore to copy rows over.

RichardJCai reviewed (commented):
Other than minor comments, privileges stuff LGTM

dt commented on pkg/ccl/backupccl/restore_planning.go:
```diff
@@ -158,6 +161,242 @@ func synthesizePGTempSchema(
+	// We bump the DescIDSeqKey to an ID higher than anything in the backup to
+	// ensure that the tables we create below won't conflict with anything in the
+	// backup.
+	if err := bumpDescIDSequenceKey(ctx, p, databasesByID, schemasByID, tablesByID, typesByID); err != nil {
```
I'm trying to piece together how this call interacts with the earlier call our caller made to bump it, before calling this function, if it was in the cluster restore case... after _that_ call we then inc it to get the tmp sys db ID, so what is this call going to bump it _to_?

adityamaru replied:
Discussed offline, because we take the max of the descriptor IDs in the backup, and the restoring clusters' sequence key we should be fine even if this method is called twice during restore planning.

dt commented on pkg/ccl/backupccl/restore_planning.go:
```diff
@@ -158,6 +161,242 @@ func synthesizePGTempSchema(
+) (map[descpb.ID]catalog.TableDescriptor, error) {
+	systemTablesWithConflictingIDs := make(map[descpb.ID]catalog.TableDescriptor)
+	err := sql.DescsTxn(ctx, p.ExecCfg(), func(ctx context.Context, txn *kv.Txn, col *descs.Collection) error {
+		systemTableDescs, err := col.GetAllTableDescriptorsInDatabase(ctx, txn, systemschema.SystemDB.GetID())
```
I still sorta wonder if we should just be looking for _any_ desc that has an ID in our maps, rather than just system descs, but I'll grant that it is the same in practice for 22.1 and we expect to not need this for 22.2 anyway, so maybe moot.

Ditto on getting just _table_ descriptors vs any descriptor that has a matching ID, but again, maybe moot since right now they're all tables anyway.

adityamaru replied:
I have a TODO after the call to `moveConflictingSystemTables` to do a sanity check that none of the descriptors in the `system.descriptors` table conflict with our backed up descriptors. In which case we should just fail the restore.

But yeah if we're getting rid of all this logic imminently then I'd rather not deal with moving types, schemas, databases at the moment. Does that sound reasonable?

dt replied:
yeah, I think this is fine to punt.

dt commented on pkg/ccl/backupccl/restore_planning.go:
```diff
@@ -158,6 +161,242 @@ func synthesizePGTempSchema(
+			return err
+		}
+
+		// ClearRange on the table span for the old system table that we've just migrated.
```
Should we pull the rest of this block out of the txn, i.e. wait until after the copy/drop/rename/etc commits before we go and clear range since clear range is non-txn'l and thus would block a txn retry?

adityamaru replied:
good catch, done.

adityamaru commented:
friendly ping @dt @RichardJCai, I want to try and catch the next 22.1 dot release

RichardJCai reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

adityamaru commented:
TFTRs! Just rebased to pickup the fixes for `TestWorkload`. Will bors on green.

adityamaru commented:
logic test timeout

bors r+
